xapian-letor/tests/api_letor.cc

```diff
@@ -143,6 +317,92 @@ DEFINE_TESTCASE(preparetrainingfile, generated)
     return true;
 }
 
+
+DEFINE_TESTCASE(preparetrainingfilethree_missingQrel, generated)
+{
+    string db_path = get_database_path("apitest_listnet_ranker3	",
+				       db_index_three_documents);
+    string data_directory = test_driver::get_srcdir() + "/testdata/";
+    string query = data_directory + "querythree.txt";
+    string qrel = data_directory + "qrelthree_onemiss.txt";
+    string training_data = data_directory + "training_data_three_onemiss.txt";
+    Xapian::prepare_training_file(db_path, query, qrel, 10,
+				  "training_output_three_miss.txt");
+    ifstream if1(training_data);
+    ifstream if2("training_output_three_miss.txt");
+    string file1;
+    string file2;
+    while (getline(if1, file1)) {
+	TEST(getline(if2, file2));
+	istringstream iss1(file1);
+	istringstream iss2(file2);
+	string temp1;
+	string temp2;
+	int i = 0;
+	while ((iss1 >> temp1) && (iss2 >> temp2)) {
+	    if (i == 0 || i == 1 || i == 21) {
+		TEST_EQUAL(temp1, temp2);
+	    } else {
+		size_t t1 = temp1.find_first_of(':');
+		size_t t2 = temp2.find_first_of(':');
+		TEST(abs(stod(temp1.substr(t1 + 1)) -
+			 stod(temp2.substr(t2 + 1))) <
+		     0.001);
+	    }
+	    i++;
+	}
+	TEST_REL(i, ==, 22);
+	TEST(!(iss2 >> temp2));
+    }
+    TEST(!getline(if2, file2));
+
```

Comment:
You should follow one guideline. In xapian-core there is no blank linke before the return statement so you should avoid adding blank lines before the return statement. Do not follow xapian-letor for coding standards as it is not formatted properly. You should ask @ojwb about more on this. **Similarly look for return statements in all the tests that you have added and make sure that you have removed the blank lines before return.**